### PR TITLE
8233760: Result of BigDecimal.toString throws overflow exception on new BigDecimal(str)

### PR DESCRIPTION
**Pull request by rgiulietti**
BigDecimal(String) currently fails to accept some strings produced by BigDecimal.toString(). This PR removes this limitation.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8233760](https://bugs.openjdk.org/browse/JDK-8233760): Result of BigDecimal.toString throws overflow exception on new BigDecimal(str)
 * [JDK-8287376](https://bugs.openjdk.org/browse/JDK-8287376): BigDecimal(String) must allow a larger range for the exponent (**CSR**)

### Reviewers
 * [Joe Darcy](https://openjdk.java.net/census#darcy) (@jddarcy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8905/head:pull/8905` \
`$ git checkout pull/8905`

Update a local copy of the PR: \
`$ git checkout pull/8905` \
`$ git pull https://git.openjdk.java.net/jdk pull/8905/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8905`

View PR using the GUI difftool: \
`$ git pr show -t 8905`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8905.diff">https://git.openjdk.java.net/jdk/pull/8905.diff</a>

</details>

### CODE REVIEW

File: test/jdk/java/math/BigDecimal/StringConstructor.java [hunk 1]
```diff
@@ -66,7 +66,6 @@ public static void main(String[] args) throws Exception {
         constructWithError("10e"+Integer.MIN_VALUE);
         constructWithError("0.01e"+Integer.MIN_VALUE);
         constructWithError("1e"+((long)Integer.MIN_VALUE-1));
-        constructWithError("1e"+((long)Integer.MAX_VALUE + 1));
```
Comment (jddarcy) [inline test/jdk/java/math/BigDecimal/StringConstructor.java:69]: Please add some test cases to demonstrate that the round-tripping that previous did not work is functional after the fix.

Comment (rgiulietti) [reply test/jdk/java/math/BigDecimal/StringConstructor.java:69]: Tests were added in the new commit.

**Comment by rgiulietti** [comment]
Hello,

here's an explanation of the changes in `BigDecimal`.

Rather than keeping track of a local `int scl` (scale) and a local `long exp` (exponent), the change only keeps track of a `long scl`, adapting it accordingly. It checks the range of `scl` before initializing the fields. This makes for a simpler code.

In addition, I couldn't resist changing C-style arrays to Java style over the whole class.

**Review by jddarcy — APPROVED** [review]

**Pull request closed without merging**